proxy/healthy_endpoints_test.go

```diff
@@ -99,107 +102,106 @@ func TestPHCForSingleHealthyEndpoint(t *testing.T) {
+	for _, algorithm := range []string{"random", "consistentHash", "roundRobin", "powerOfRandomNChoices"} {
+		t.Run(algorithm, func(t *testing.T) {
+			services := []*httptest.Server{}
+			for i := 0; i < 3; i++ {
+				service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+				}))
+				services = append(services, service)
+				defer service.Close()
+			}
+			endpointRegistry := defaultEndpointRegistry()
+
+			doc := fmt.Sprintf(`* -> backendTimeout("5ms") -> consistentHashKey("${request.header.ConsistentHashKey}") -> <%s, "%s", "%s", "%s">`,
+				algorithm, services[0].URL, services[1].URL, services[2].URL)
+			tp, err := newTestProxyWithParams(doc, Params{
+				EnablePassiveHealthCheck: true,
+				EndpointRegistry:         endpointRegistry,
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer tp.close()
+
+			ps := httptest.NewServer(tp.proxy)
+			defer ps.Close()
+
+			failedReqs := 0
+			for i := 0; i < nRequests; i++ {
+				req, err := http.NewRequest("GET", ps.URL, nil)
+				if err != nil {
+					t.Fatal(err)
+				}
+				req.Header.Add("ConsistentHashKey", fmt.Sprintf("%d", i))
+
+				rsp, err := ps.Client().Do(req)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if rsp.StatusCode != http.StatusOK {
+					failedReqs++
+				}
+				rsp.Body.Close()
+			}
+			assert.Equal(t, 0, failedReqs)
+		})
 	}
 }
 
 func TestPHCForMultipleHealthyAndOneUnhealthyEndpoints(t *testing.T) {
-	services := []*httptest.Server{}
-	for i := 0; i < 3; i++ {
-		service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			w.WriteHeader(http.StatusOK)
-		}))
-		services = append(services, service)
-		defer service.Close()
-	}
-	endpointRegistry := defaultEndpointRegistry()
-
-	doc := fmt.Sprintf(`* -> <random, "%s", "%s", "%s">`, services[0].URL, services[1].URL, services[2].URL)
-	tp, err := newTestProxyWithParams(doc, Params{
-		EnablePassiveHealthCheck:   true,
-		EndpointRegistry:           endpointRegistry,
-		CustomHttpRoundTripperWrap: newRoundTripperUnhealthyHost(&RoundTripperUnhealthyHostOptions{Host: services[0].URL[7:], Probability: rtFailureProbability}),
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer tp.close()
-
-	ps := httptest.NewServer(tp.proxy)
-	defer ps.Close()
-
-	failedReqs := 0
-	for i := 0; i < nRequests; i++ {
-		rsp, err := ps.Client().Get(ps.URL)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		if rsp.StatusCode != http.StatusOK {
-			failedReqs++
-		}
-		rsp.Body.Close()
+	for _, algorithm := range []string{"random", "consistentHash", "roundRobin", "powerOfRandomNChoices"} {
+		t.Run(algorithm, func(t *testing.T) {
+			services := []*httptest.Server{}
+			for i := 0; i < 3; i++ {
+				serviceNum := i
+				service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					if serviceNum == 0 {
+						// emulating unhealthy endpoint
+						time.Sleep(15 * time.Millisecond)
```

Comment:
Make it 10-20x of the backendTimout argument just in case.

Reply:
Increased time.Sleep to 100ms.